[WIP] Move XNNPACKQuantizer from PyTorch to ExecuTorch

digantdesai opened the pull request:
Summary:
This replicates XNNPACKQuantizer from PyTorch to ExecuTorch.

Rationale:
Main motivation is to avoid pytorch pin update in OSS after updating XNNPACKQuantizer, which can be rather frequent.
Other impact and considerations:
PT2e flow (which lives in PyTorch) relies havily on XNNPACKQuantizer for a "example" implementation for quantizer and more importantly tests. Fow now, we will keep the torch.ao.quantization.xnnpack_quantizer as is but mark is as not BC, and deprecated to discourace future new dependencies on it.
Other OSS repository using XNNPACKQuantizer from PyTorch now have to take an additional dependency on ExecuTorch.

Differential Revision: D68191752


cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jerryzh168 commented on test/quantization/pt2e/test_representation.py:
cc @kimishpatel feels like reference representation is not really used, should we delete?

kimishpatel replied:
WHy is this test being deleted entirely though? I am guessing because xnnpack quantizer is moving hence? I would have to think a bit on removing reference representation, and we should discuss. Potentially we remove those in favor providing an API that will use customer defined representation

jerryzh168 replied:
yeah this test is moved to executorch I think. sounds good, custom defined representation makes sense as well

digantdesai replied:
I am thinking of adding a `dummy_quantizer` in the test dir to run these tests especially qat ones.

jerryzh168 commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -31,7 +31,7 @@ def generate_numeric_debug_handle(ep: ExportedProgram) -> None:
 
         m = ep.module()
-        quantizer = XNNPACKQuantizer()
+        quantizer = XNNPACKQuantizer() # In the ExecuTorch repo.
```
please add an import for this

digantdesai replied:
what do you mean? This is an example, do you mean to put import line in addition to the comment in the doc string?

jerryzh168 replied:
yeah.. just to make it clearer since it's in executorch now

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -64,6 +58,73 @@
 
+# In the absence of better name, just winging it with QuantizationConfig
+@dataclass(eq=True, frozen=True)
+class QuantizationConfig:
```
we can probably put these in quantizer/utils.py since these are shared, but should be clear that there is no requirement of quantizers using them

digantdesai replied:
I can move them, but IIRC no one else is using it so kept it here.

albanD commented on torch/ao/quantization/quantizer/embedding_quantizer.py:
```diff
@@ -12,18 +13,38 @@
 ]
 
+# In the absence of better name, just winging it with QuantizationConfig
+@dataclass(eq=True, frozen=True)
```
These need to be added to the `__all__` block above and be properly documented on the doc website if you want to move them here as public API.

albanD reviewed (commented):
Simplification!!
skipping pr sanity check is obviously ok for this one!

kimishpatel commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -236,6 +236,12 @@ def not_module_type_or_name_filter(n: Node) -> bool:
 
 
 class XNNPACKQuantizer(Quantizer):
+    """
```
There are some decorator that allow you to declare either not bc compatible or something along the lines. @jerryzh168 do you know of any that can be used here?

jerryzh168 replied:
[this?](https://github.com/pytorch/pytorch/blob/adbbcd87d9fafe0aa34c69dd8f780f919a24461c/torch/fx/proxy.py#L40)

we can do that, looks like it modifies docstring: https://github.com/pytorch/pytorch/blob/adbbcd87d9fafe0aa34c69dd8f780f919a24461c/torch/fx/_compatibility.py#L12 might be useful if we mark it in a lot of things

digantdesai replied:
Ok doing (1) for now, *and* keeping this docstring.

jerryzh168 reviewed (approved):
please update the PR summary as well about the migration plan

digantdesai committed:
Move XNNPACKQuantizer from PyTorch to ExecuTorch (#144940)

Summary:
X-link: https://github.com/pytorch/executorch/pull/7804

Pull Request resolved: https://github.com/pytorch/pytorch/pull/144940

This migrates XNNPACKQuantizer from PyTorch to ExecuTorch.

Rationale:
Main motivation is to avoid pytorch pin update in OSS after updating XNNPACKQuantizer, which can be rather frequent.

Other impact and considerations:
- PT2e flow (which lives in PyTorch) relies havily on XNNPACKQuantizer for a "example" implementation for quantizer and more importantly tests. Fow now, we will keep the torch.ao.quantization.xnnpack_quantizer as is but mark is as not BC, and deprecated to discourace future new dependencies on it.

- Other OSS repository using XNNPACKQuantizer from PyTorch now have to take an additional dependency on ExecuTorch.

Test Plan: CI

Reviewed By: mcr229

Differential Revision: D68191752
